### PR TITLE
Fix GLSL failures due to fix of #3221

### LINKED ISSUE #3371: apps/glsl is failing
$ make distrib && cd apps/glsl && make run
Internal error at CodeGen_GPU_Host.cpp:54 triggered by user code at ../../distrib/tools/GenGen.cpp:4:
Condition failed: is_zero(op->min)

### PR DESCRIPTION
**Pull request by shoaibkamil**
Moving `canonicalize_gpu_vars` after storage flattening causes OpenGL to fail, since it relies on some transformations done during storage flattening.  This PR tries to resolve #3371 while preserving the fix to #3221 by moving canonicalization just before storage flattening.

### CODE REVIEW

**Comment by steven-johnson** [comment]
LGTM, but review from @abadams essential before landing

**Comment by steven-johnson** [comment]
The arm32 build is failing with:
```
FAILED TEST: gpu_bounds_inference_failure
LLVM ERROR: Program used external function '__sync_val_compare_and_swap_4' which could not be resolved!
```

**Comment by abadams** [comment]
Looks like the test unconditionally compiles something for host-cuda, which isn't going to work on every platform.

**Comment by abadams** [comment]
Probably better to get_jit_target_from_environment and skip the test if cuda is not enabled

**Comment by shoaibkamil** [comment]
Ah, doh.  Will fix.

File: test/correctness/gpu_bounds_inference_failure.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+#include "Halide.h"
+
+using namespace Halide;
+
+int main(int argc, char *argv[]) {
+
+  if (!t.has_feature(Target::CUDA)) {
+    printf("Not running test because cuda not enabled\n");
+  }
```
Comment (abadams) [inline test/correctness/gpu_bounds_inference_failure.cpp]: Erm, I think you also need to return :)

Also 4-space indentation.

**Pull request merged**